### ISSUE TITLE
icp setIndices doesn't work

### ISSUE THREAD

**Issue by yzl96** [PointCloudLibrary/pcl#3109]
## Your Environment

* Operating System and version:
* Compiler:
* PCL Version:

## Code to Reproduce

`pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_in (new pcl::PointCloud<pcl::PointXYZ>);
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_out (new pcl::PointCloud<pcl::PointXYZ>);

    // Fill in the CloudIn data
    cloud_in->width    = 1000;
    cloud_in->height   = 1;
    cloud_in->is_dense = false;
    cloud_in->points.resize (cloud_in->width * cloud_in->height);
    for (size_t i = 0; i < cloud_in->points.size (); ++i)
    {
        cloud_in->points[i].x = 1024 * rand () / (RAND_MAX + 1.0f);
        cloud_in->points[i].y = 1024 * rand () / (RAND_MAX + 1.0f);
        cloud_in->points[i].z = 1024 * rand () / (RAND_MAX + 1.0f);
    }

    cloud_out->width    = 100;
    cloud_out->height   = 1;
    cloud_out->is_dense = false;
    cloud_out->points.resize (cloud_out->width * cloud_out->height);

    for (size_t i = 0; i < cloud_out->points.size (); ++i)
    {
        cloud_out->points[i].x = cloud_in->points[i].x + 0.7f;
        cloud_out->points[i].y = cloud_in->points[i].y;
        cloud_out->points[i].z = cloud_in->points[i].z;
    }
    std::cout << cloud_out->points.size() << std::endl;

    boost::shared_ptr< std::vector<int> > pr(new std::vector<int>);
   // first try
   //   for(int i=0; i<100; i+=1)
    //    pr->push_back(i);
 // second try
    for(int i=100; i<200; i+=1)
        pr->push_back(i);

    pcl::IterativeClosestPoint<pcl::PointXYZ, pcl::PointXYZ> icp;
    icp.setMaxCorrespondenceDistance(0.5);
    icp.setTransformationEpsilon(1e-12);
    icp.setEuclideanFitnessEpsilon(0.0001);
    icp.setMaximumIterations (1000000);
    icp.setInputSource(cloud_in);
    icp.setInputTarget(cloud_out);

    //icp.setIndices(pr);
    pcl::PointCloud<pcl::PointXYZ> Final;
    icp.align(Final);

    std::cout << "has converged:" << icp.hasConverged() << " score: " <<
              icp.getFitnessScore() << std::endl;

    std::cout << icp.getFinalTransformation() << std::endl;

    pcl::transformPointCloud(*cloud_out,*cloud_out,icp.getFinalTransformation().inverse());`

**Comment by taketwo** [comment]
Yes, you use different points in the input cloud, but the transform is nevertheless the same, `0.7` along the _x_ axis. Why would it be different?

**Comment by yzl96** [comment]
@taketwo , I did use different points in the input cloud, but my output point cloud is fixed, my output cloud have 100 points, which are generated by add 0.7 to x axis from first 100 points of the input point cloud, if I set indices [0,99] of input point cloud, there is a transformation 0.7 in x  axis between input cloud subset and output cloud, but if I set indices [100,199], there should no relation between input cloud subset and output cloud, my output point cloud is generate from the first 100 points of input cloud, it has no relation with the input cloud subset (indices from 100 to 199).

**Comment by yzl96** [comment]
@taketwo , how can it be same, the setIndices function choose the subset of input point cloud to register with output cloud, I use different subset of input cloud, and output point cloud is fixed, why the transfomation is the same, the input of registration is different.

**Comment by taketwo** [comment]
Hi, actually you may be right about this one. I had a brief look at the ICP code and it does not seem to pay attention to supplied indices indeed.

**Comment by AlrIsmail** [comment]
While implementing the fix for the registration indices, I noticed an additional detail regarding the getFitnessScore() method in pcl::Registration.

Even when setIndices is used, getFitnessScore() currently iterates over the entire input cloud to calculate the final fitness score. This means the reported score reflects the alignment of the full dataset, not just the specified subset.

In my current PR, I have focused strictly on the registration transformation (ensuring the actual alignment math respects the indices), as this was the primary bug reported. I have left the getFitnessScore behavior unchanged to keep the PR scope focused, but I wanted to raise this for discussion: Should getFitnessScore also be capped to the provided indices in a follow-up task to ensure API consistency?

**Comment by larshg** [comment]
> While implementing the fix for the registration indices, I noticed an additional detail regarding the getFitnessScore() method in pcl::Registration.
> 
> Even when setIndices is used, getFitnessScore() currently iterates over the entire input cloud to calculate the final fitness score. This means the reported score reflects the alignment of the full dataset, not just the specified subset.
> 
> In my current PR, I have focused strictly on the registration transformation (ensuring the actual alignment math respects the indices), as this was the primary bug reported. I have left the getFitnessScore behavior unchanged to keep the PR scope focused, but I wanted to raise this for discussion: Should getFitnessScore also be capped to the provided indices in a follow-up task to ensure API consistency?

Lets open a new issue for that - so it can be discussed separately. I guess it would be nice if one could toggle whether they want the fitnessScore calculated on entire clouds or just the selected indices as well. But open a new issue and we can continue the discussion there.
